src/lib/api/prefetch.ts

```diff
@@ -0,0 +1,128 @@
+import { PickerOptions } from './../picker';
+// import { requestWithSource } from '../api/request';
+import { FsRequest } from '../request';
+import { cloneDeep as lodashCloneDeep, merge as lodashMerge } from 'lodash';
+
+export type PrefetchOptionsSetting = {
+  inapp_browser?: boolean;
+  customsource?: boolean;
+};
+
+export type PrefetchOptionsPermissions = {
+  intelligent_ingestion?: boolean;
+  blocked?: boolean;
+  blacklisted?: boolean;
+  whitelabel?: boolean;
+  transforms_ui?: boolean;
+};
+
+export enum PrefetchOptionsEvents {
+  PICKER = 'picker',
+  TRANSFORM_UI = 'transform_ui',
+}
+
+type PrefetchOptions = {
+  pickerOptions?: PickerOptions;
+  settings?: PrefetchOptionsSetting;
+  permissions?: PrefetchOptionsPermissions;
+  events?: PrefetchOptionsEvents[];
+};
+
+interface PrefetchRequest {
+  apikey: string;
+  security?: {
+    policy?: string;
+    signature?: string;
+  };
+  permissions?: string[];
+  settings?: string[];
+  events?: PrefetchOptionsEvents[];
+  picker_config?: PickerOptions;
+}
+
+type PrefetchResponse = {
+  blocked?: boolean;
+  settings?: PrefetchOptionsSetting;
+  permissions?: PrefetchOptionsPermissions;
+  updated_config?: {
+    fromSources?: string[];
+  };
+};
+
+/**
+ * @private
+ */
+export class Prefetch {
+  private session: Session;
+  private prefetchUrl: string;
+  private configToCheck: PickerOptions;
+
+  constructor(session: Session) {
+    this.session = session;
+    this.prefetchUrl = session.urls.uploadApiUrl;
+  }
+
+  async getConfig({ pickerOptions, settings, permissions, events }: PrefetchOptions) {
+    if (this.session.prefetch) {
+      await FsRequest.post(`${this.prefetchUrl}/prefetch`, { events });
+      return Promise.resolve(this.session.prefetch);
+    }
+
+    const configToSend = this.cleanUpCallback(pickerOptions);
+    const permissionsKeys = [...Object.keys(permissions)];
+
+    let paramsToSend: PrefetchRequest = {
+      apikey: this.session.apikey,
+      permissions: permissionsKeys,
+      settings: Object.keys(settings),
+      picker_config: configToSend,
+      events,
+    };
+
+    if (this.session.policy && this.session.signature) {
+      paramsToSend.security = { policy: this.session.policy, signature: this.session.signature };
+    }
+
+    const response = await FsRequest.post(`${this.prefetchUrl}/prefetch`, paramsToSend).then(res => res.data);
+
+    // const response = await requestWithSource()
+    //   .post(`${this.prefetchUrl}/prefetch`, paramsToSend)
+    //   .then(res => res.data);
+
+    return this.reassignCallbacks(response);
+  }
+
+  private cleanUpCallback(pickerOptions: PickerOptions) {
+    this.configToCheck = lodashCloneDeep(pickerOptions);
+
+    Object.keys(this.configToCheck).map(key => {
+      if (key.indexOf('on') === 0) {
+        this.configToCheck[key] = undefined;
+      }
+    });
+
+    return this.configToCheck;
+  }
+
+  private reassignCallbacks(response: PrefetchResponse) {
+    return lodashMerge({}, response);
```

Comment:
Youy are merging response with empty object instead of old config, so it will not reassign old callbacks